src/software/new_geom/vector.h

```diff
@@ -0,0 +1,521 @@
+ *
+ * @paramp u the Vector to substract from
+ * @param v the Vector to subtract
+ *
+ * @return the new value of Vector u
+ */
+Vector &operator-=(Vector &u, const Vector &v);
+
+/**
+ * Multiplies a vector by a scalar
+ *
+ * @param s the scaling factor
+ * @param v the vector to scale
+ *
+ * @return the scaled vector
+ */
+constexpr Vector operator*(double s, const Vector &v);
+
+/**
+ * Multiplies a vector by a scalar
+ *
+ * @param p the vector to scale
+ * @param s the scaling factor
+ *
+ * @return the scaled vector
+ */
+constexpr Vector operator*(const Vector &p, double s);
+
+/**
+ * Scales a vector by a scalar
+ *
+ * @param p the vector to scale
+ * @param s the scaling factor
+ *
+ * @return p scaled by the scaling factor
+ */
+Vector &operator*=(Vector &p, double s);
+
+/**
+ * Divides a vector by a scalar
+ *
+ * @param p the vector to scale
+ * @param s the scalar to divide by
+ *
+ * @return the scaled vector
+ */
+constexpr Vector operator/(const Vector &p, double s);
+
+/**
+ * Scales a vector by a scalar
+ *
+ * @param p the vector to scale
+ * @param s the sclaing factor
+ *
+ * @return p scaled by the scaling factor
+ */
+Vector &operator/=(Vector &p, double s);
+
+/**
+ * Prints a vector to a stream
+ *
+ * @param os the stream to print to
+ * @param v the Vector to print
+ *
+ * @return the stream with the Vector printed
+ */
+inline std::ostream &operator<<(std::ostream &os, const Vector &v);
+
+/**
+ * Compares two Vectors for equality
+ *
+ * @param u the first Vector
+ * @param v the second Vector
+ *
+ * @return true if the two vectors represent the same vector, and false otherwise
+ */
+constexpr bool operator==(const Vector &u, const Vector &v);
+
+/**
+ * Compares two vectors for inequality
+ *
+ * @param u the first Vector
+ * @param v the second Vector
+ *
+ * @return true if the two vectors represent different vectors, and false otherwise
+ */
+constexpr bool operator!=(const Vector &u, const Vector &v);
+
+inline Vector Vector::createFromAngle(const Angle &angle)
+{
+    return Vector(angle.cos(), angle.sin());
+}
+
+inline constexpr Vector::Vector() : _x(0.0), _y(0.0) {}
+
+inline constexpr Vector::Vector(double x, double y) : _x(x), _y(y) {}
+
+inline constexpr Vector::Vector(const Vector &v) : _x(v.x()), _y(v.y()) {}
+
+inline constexpr double Vector::x() const
+{
+    return _x;
+}
+
+inline constexpr double Vector::y() const
+{
+    return _y;
+}
+
+inline void Vector::set(double x, double y)
+{
+    this->_x = x;
+    this->_y = y;
+}
+
+inline void Vector::setX(double x)
+{
+    this->_x = x;
+}
+
+inline void Vector::setY(double y)
+{
+    this->_y = y;
+}
+
+inline constexpr double Vector::lensq() const
+{
+    return _x * _x + _y * _y;
+}
+
+inline double Vector::len() const
+{
+    return std::hypot(_x, _y);
+}
+
+inline Vector Vector::norm() const
+{
+    return len() < 1.0e-9 ? Vector() : Vector(_x / len(), _y / len());
```

Comment:
Why are we rounding down like this?

Reply:
I suspect it's from the old implementation. I can't think of any reason other than numerical stability of some kind? Otherwise you could end up with "infinitely small" vectors? which if you're doing operation on the length could cause values to blow up?

Reply:
I mean, a zero length vector would have the same effect as an "infinitely small" one. If we were really in that situation, I'd say it's the job of the user to compare the vector norm to `EPSILON` (but definitely not `NEAR`)